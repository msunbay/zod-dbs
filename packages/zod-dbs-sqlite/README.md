# zod-dbs-sqlite

SQLite provider for [zod-dbs](https://github.com/msunbay/zod-dbs).

## Installation

```bash
npm install zod-dbs-sqlite better-sqlite3
```

> Note: This package has a peer dependency on `better-sqlite3`. If you don't already have it, install it alongside.

## Cli Usage

```bash
npm install zod-dbs-cli zod-dbs-sqlite better-sqlite3

npx zod-dbs --provider sqlite
```

## Programmatic Usage

```bash
npm install zod-dbs zod-dbs-sqlite better-sqlite3
```

```ts
import { generateZodSchemas } from 'zod-dbs';
import { SqliteProvider } from 'zod-dbs-sqlite';

const provider = new SqliteProvider();

await generateZodSchemas({
  provider,
  config: {
    // See zod-dbs documentation for available options
    // Example (SQLite):
    // database: 'path/to/db.sqlite',
    // schemaName: 'main',
    // outputDir: './zod-schemas',
  },
});
```
