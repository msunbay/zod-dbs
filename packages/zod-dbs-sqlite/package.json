{
  "name": "zod-dbs-sqlite",
  "version": "1.0.1",
  "description": "SQLite provider for zod-dbs",
  "keywords": [
    "zod",
    "zod-dbs",
    "typescript",
    "database",
    "schema",
    "sqlite",
    "sqlite3",
    "embedded",
    "type-safe",
    "validation",
    "sql",
    "better-sqlite3",
    "introspection",
    "schema-generator"
  ],
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "type": "module",
  "scripts": {
    "build": "rm -rf ./dist && tsc --project tsconfig.build.json",
    "test:unit": "vitest --project unit run",
    "test:types": "tsc --noEmit"
  },
  "author": "msunbay",
  "license": "ISC",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/msunbay/zod-dbs.git",
    "directory": "packages/zod-dbs-sqlite"
  },
  "publishConfig": {
    "access": "public"
  },
  "files": [
    "dist"
  ],
  "dependencies": {
    "zod-dbs-core": "workspace:*"
  },
  "devDependencies": {
    "@types/better-sqlite3": "^7.6.13",
    "better-sqlite3": "^11.7.0"
  },
  "peerDependencies": {
    "better-sqlite3": ">=11"
  }
}
