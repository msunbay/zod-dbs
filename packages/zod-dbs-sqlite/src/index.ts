export { createClient } from './client.js';
export { createProvider } from './provider.js';
export { SqliteProvider } from './SqliteProvider.js';
export type * from './types.js';
