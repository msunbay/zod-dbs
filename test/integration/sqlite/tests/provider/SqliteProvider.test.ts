import { SqliteProvider } from 'zod-dbs-sqlite';

import { getProviderConfig } from '../../../utils/context.js';

it('returns raw schema column information', async () => {
  const provider = new SqliteProvider();

  const info = await provider.fetchSchemaInfo(getProviderConfig());

  expect(info).toBeDefined();
  expect(info.length).toBeGreaterThan(0);
  expect(info).toMatchSnapshot('rawColumns');
});

it('returns schema models', async () => {
  const provider = new SqliteProvider();

  const info = await provider.getSchemaInformation(getProviderConfig());

  expect(info).toBeDefined();
  expect(info.tables).toBeDefined();

  const userTable = info.tables.find((t) => t.name === 'users')!;
  expect(userTable).toBeDefined();
  expect(userTable.columns).toBeDefined();
  expect(userTable.columns.length).toBeGreaterThan(0);
  expect(userTable).toMatchSnapshot('userTable');
});

it('detects enum-like columns via CHECK IN constraints', async () => {
  const provider = new SqliteProvider();

  const info = await provider.getSchemaInformation(getProviderConfig());

  const userTable = info.tables.find((t) => t.name === 'users')!;
  const statusCol = userTable.columns.find((c) => c.name === 'status');

  expect(statusCol?.isEnum).toBe(true);
  expect(statusCol?.enumValues).toEqual(['active', 'inactive', 'banned']);
});
