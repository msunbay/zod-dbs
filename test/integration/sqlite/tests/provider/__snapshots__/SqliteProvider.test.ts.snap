// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns raw schema column information > rawColumns 1`] = `
[
  {
    "dataType": "integer",
    "defaultValue": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": true,
    "maxLen": undefined,
    "name": "id",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "text",
    "defaultValue": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": false,
    "isSerial": false,
    "maxLen": undefined,
    "name": "name",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "text",
    "defaultValue": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "email",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "text",
    "defaultValue": "'active'",
    "enumValues": [
      "active",
      "inactive",
      "banned",
    ],
    "isArray": false,
    "isEnum": true,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "status",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "json",
    "defaultValue": undefined,
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "profile",
    "tableName": "users",
    "tableType": "table",
  },
  {
    "dataType": "text",
    "defaultValue": "datetime('now')",
    "isArray": false,
    "isEnum": false,
    "isNullable": true,
    "isSerial": false,
    "maxLen": undefined,
    "name": "created_at",
    "tableName": "users",
    "tableType": "table",
  },
]
`;

exports[`returns schema models > userTable 1`] = `
{
  "columns": [
    {
      "dataType": "integer",
      "defaultValue": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": true,
      "isWritable": false,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "id",
      "tableName": "users",
      "tableType": "table",
      "zodType": "int",
    },
    {
      "dataType": "text",
      "defaultValue": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": false,
      "isReadOptional": false,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": false,
      "maxLen": undefined,
      "name": "name",
      "tableName": "users",
      "tableType": "table",
      "zodType": "string",
    },
    {
      "dataType": "text",
      "defaultValue": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "email",
      "tableName": "users",
      "tableType": "table",
      "zodType": "string",
    },
    {
      "dataType": "text",
      "defaultValue": "'active'",
      "enumValues": [
        "active",
        "inactive",
        "banned",
      ],
      "isArray": false,
      "isEnum": true,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "status",
      "tableName": "users",
      "tableType": "table",
      "zodType": "string",
    },
    {
      "dataType": "json",
      "defaultValue": undefined,
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "profile",
      "tableName": "users",
      "tableType": "table",
      "zodType": "json",
    },
    {
      "dataType": "text",
      "defaultValue": "datetime('now')",
      "isArray": false,
      "isEnum": false,
      "isNullable": true,
      "isReadOptional": true,
      "isSerial": false,
      "isWritable": true,
      "isWriteOptional": true,
      "maxLen": undefined,
      "name": "created_at",
      "tableName": "users",
      "tableType": "table",
      "zodType": "string",
    },
  ],
  "name": "users",
  "schemaName": undefined,
  "type": "table",
}
`;
